File: ocaml/day6/main.ml
open Core

let datastream =
  In_channel.read_lines "../inputs/day6.txt" |> List.hd_exn |> String.to_list

let first_marker_pos data marker_size =
  let is_marker marker =
    List.length marker = marker_size
    && not (List.contains_dup ~compare:compare_char marker)
  in
  let rec iter_markers xs pos =
    match xs with
    | c :: c' :: cs' ->
        if List.length cs' >= marker_size - 2 then
          if is_marker (c :: c' :: List.take cs' (marker_size - 2)) then
            pos + marker_size
          else iter_markers (c' :: cs') (pos + 1)
        else -1
    | _ -> -1
  in
  iter_markers data 0

let () =
  printf "Task 1: %d\n" @@ first_marker_pos datastream 4;
  printf "Task 2: %d\n" @@ first_marker_pos datastream 14

File: rust/day6/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Result as IOResult};

const INPUT_PATH: &str = "../inputs/day6.txt";

const START_OF_PACKET_MARKER_SIZE: usize = 4;
const START_OF_MESSAGE_MARKER_SIZE: usize = 14;

fn read_datastream() -> IOResult<String> {
    Ok(BufReader::new(File::open(INPUT_PATH)?)
        .lines()
        .flat_map(|line| line)
        .next()
        .expect("Input must contain a line"))
}

fn first_marker_pos(datastream: &str, marker_size: usize) -> usize {
    let (pos, _) = datastream
        .as_bytes()
        .windows(marker_size)
        .enumerate()
        .find(|(_, window)| !(1..marker_size).any(|i| window[i..].contains(&window[i - 1])))
        .expect("There's an answer to this problem ... right?");
    pos + marker_size
}

fn main() -> IOResult<()> {
    let datastream = read_datastream()?;
    println!(
        "Task 1: {}",
        first_marker_pos(&datastream, START_OF_PACKET_MARKER_SIZE)
    );
    println!(
        "Task 2: {}",
        first_marker_pos(&datastream, START_OF_MESSAGE_MARKER_SIZE)
    );
    Ok(())
}
